### PR TITLE
transformations: add memref-stream-interleave

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

**Review by nazavode — APPROVED** [review]
Beautiful.

File: xdsl/transforms/memref_stream_interleave.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                        )
+                        + (
+                            AffineExpr.dimension(interleave_bound_index)
+                            * self.pipeline_depth
+                            + AffineExpr.dimension(m.data.num_dims),
+                        )
+                        + tuple(
+                            AffineExpr.dimension(i)
+                            for i in range(
+                                interleave_bound_index + 1, m.data.num_dims + 2
```
Comment (georgebisbas) [inline xdsl/transforms/memref_stream_interleave.py:102]: Where do these numbers stem from?

Comment (superlopuh) [reply xdsl/transforms/memref_stream_interleave.py:102]: The idea is that there is an extra dimension added somewhere in the middle of the affine map domain. This is the slice of the domain after the updated value, so it's going from `inserted_index + 1 to old_len + 1` inclusive.

**Comment by superlopuh** [comment]
Lowering works OK, let's merge this and keep going.

**Pull request merged**